components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F

```diff
@@ -0,0 +1,380 @@
+
+         do k = maxLevelEdgeTop(iEdge), minLevelEdgeBot(iEdge), -1
+            heightEdgeNow(k, iEdge) = &
+                           0.5_RKIND * (heightCellNow(k, iCell) + heightCellNow(k, jCell))
+            heightEdgeNew(k, iEdge) = &
+                           0.5_RKIND * (heightCellNew(k, iCell) + heightCellNew(k, jCell))
+         end do
+      end do
+
+   !----------------------------------------------------------------------------
+   !- ACTUAL REMAPPING FROM HERE
+
+   !- just simple hard-coded options for PPR at this stage...
+
+      call mpas_log_write('-- remap velocity')
+      if (associated(highFreqThickness) .and. associated(lowFreqDivergence)) then
+         nVars = 3 ! how many tracers to be remapped concurrently per column
+      else
+         nVars = 1 ! how many tracers to be remapped concurrently per column
+      endif
+      nDoFs = 1 ! always 1 for MPAS, it'd be >1 for DG methods (with more DoF per layer)
+
+      allocate(dataNow(nDoFs, nVars, nVertLevels))
+      allocate(dataNew(nDoFs, nVars, nVertLevels))
+      allocate(bcUpper(nVars))
+      allocate(bcLower(nVars))
+      
+      dataNow = 0.0_RKIND
+      dataNew = 0.0_RKIND
+
+      bcUpper%bcopt = bc_upper ! "loose" => extrapolate
+      bcLower%bcopt = bc_lower
+
+      call work%init(nVertLevels + 1, nVars, opts) ! this is internal workspace for PPR
+
+   !- Remap all edge-centred variables from heightEdgeNow to heightEdgeNew
+      if (associated(highFreqThickness) .and. associated(lowFreqDivergence)) then
+         do iEdge = 1, nEdges
+
+            kTop = minLevelEdgeBot(iEdge)
+            kBot = maxLevelEdgeTop(iEdge)
+
+            nLevels = kBot - kTop + 1
+
+            if (nLevels .lt. 2) cycle
+
+            dataNow(1, 1, kTop:kBot) = normalVelocity   (kTop:kBot, iEdge)
+            dataNow(1, 2, kTop:kBot) = highFreqThickness(kTop:kBot, iEdge)
+            dataNow(1, 3, kTop:kBot) = lowFreqDivergence(kTop:kBot, iEdge)
+
+            call rmap1d(nLevels, nLevels, nVars ,nDoFs, &
+               heightEdgeNow(kTop:kBot+1, iEdge), &
+               heightEdgeNew(kTop:kBot+1, iEdge), &
+               dataNow, dataNew, &
+               bcUpper, bcLower, work, opts)
+
+            normalVelocity   (kTop:kBot, iEdge) = dataNew(1, 1, kTop:kBot)
+            highFreqThickness(kTop:kBot, iEdge) = dataNew(1, 2, kTop:kBot)
+            lowFreqDivergence(kTop:kBot, iEdge) = dataNew(1, 3, kTop:kBot)
+         end do
+      else
+         do iEdge = 1, nEdges
+
+            kTop = minLevelEdgeBot(iEdge)
+            kBot = maxLevelEdgeTop(iEdge)
+
+            nLevels = kBot - kTop + 1
+
+            if (nLevels .lt. 2) cycle
+
+            dataNow(1, 1, kTop:kBot+1) = normalVelocity(kTop:kBot+1, iEdge)
+
+            call rmap1d(nLevels, nLevels, nVars ,nDoFs, &
+               heightEdgeNow(kTop:kBot+1, iEdge), &
+               heightEdgeNew(kTop:kBot+1, iEdge), &
+               dataNow, dataNew, &
+               bcUpper, bcLower, work, opts)
```

Comment:
@dengwirda @xylar, There is currently an issue with the remapping of `normalVelocity` at the bottom of the domain. 

When `nLevels = kBot - kTop + 2`, the remapped `normalVelocity(kBot-1, iEdge)` becomes a high value while `normalVelocity(kBot:, iEdge)` are zero. 

When `nLevels = kBot - kTop + 1`, the remapped `normalVelocity(kBot-1, iEdge)` looks correct but `normalVelocity(kBot, iEdge)` is zero rather than the non-zero value because we're not remapping it. 

I'm not sure why this problem does not arise for tracers. The `heightEdge*` values are as expected.

Reply:
Resolved by 8a26320